lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,236 @@
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+require 'set'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # Files present in :files are not processed further. The others
+      #   are added to the Hash. This behavior is the same implemented in openstack-infra/shade
+      #   see https://github.com/openstack-infra/shade/blob/1d16f64fbf376a956cafed1b3edd8e51ccc16f2c/shade/openstackcloud.py#L1200
+
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :template_ori
+        attr_reader :visited
+        attr_reader :max_files_size
+
+        def initialize(template, files = nil)
+          @template = template
+          @template_ori = template
+          @template_base_url = nil
+          @files = files || {}
+          @max_files_size = (128 * 1 << 10)
+          @visited = {}
+        end
+
+        def yaml_load(content)
+          return YAML.load(content) if RUBY_VERSION < "2.1"
+
+          YAML.safe_load(content, [Date])
+        end
+
+        def get_files
```

Comment:
@bzwei How can I test private methods?

Reply:
with rspec you test private method as if it is public

Reply:
@bzwei fog-openstack uses minitest, which  seems not doing the job. 

Is it worth to replace all calls with the `send` counterpart?

```
        file_resolver.send(:get_file_contents, data)
```

Otherwise I'll:

  - I'll leave as public the principal template resolution methods (eg. may be useful in further development, like adopting stacks, resource_registry, ..)
  - move private methods to a different file.rb
  - we may add or ignore tests for private methods (imho they will be useful for troubleshooting/debugging)
  
Let me know and I'll do ;)

Reply:
done.